docs/developers/listener.md

```diff
@@ -0,0 +1,193 @@
+	// do stuff
+}
+```
+
+NORMAL is the default value if you do not specify an order.
+
+### Registering listeners
+
+While your main plugin class is automatically registered, you will need to register with the EventManager any other listeners you have:
+
+```java
+server.getEventManager().register(plugin, listener);
+```
+
+Both parameters are `Object`, first the instance of your plugin, second the listener to register. For example:
+
+```java
+@Plugin(id = "myfirstplugin", name = "My Plugin", version = "0.1.0", dependencies = {@Dependency(id = "wonderplugin")})
+public class VelocityTest {
+  
+  private final ProxyServer server;
+  private final Logger logger;
+  
+  @Inject
+  public VelocityTest(ProxyServer server, Logger logger) {
+    this.server = server;
+    this.logger = logger;
+  }
+  
+  @Subscribe
+  public void onInitialize(ProxyInitializeEvent event) {
+    server.getEventManager().register(this, new MyListener());
+  }
+}
+
+public class MyListener {
+
+  @Subscribe(order = PostOrder.EARLY)
+  public void onPlayerChat(PlayerChatEvent event) {
+    // do something here
+  }
+
+}
+```
+
+### An Alternative Approach
+
+As an alternative to `@Subscribe`, you can also use the functional `EventHandler` interface and register yours with
+`register(Object plugin, Class<E> eventClass, EventHandler<E> handler)`
+
+## Creating Events
+
+Creating events on Velocity is somewhat different than on other platforms. However, it is very similar for the most part.
+
+### Creating the Event Class
+
+First we need to create a class for our event. In this tutorial we'll assume you're making a private messaging plugin, and thus use
+a `PrivateMessageEvent`. Most of this part is boilerplate.
+
+```java
+public class PrivateMessageEvent {
+
+  private final Player sender;
+  private final Player recipient;
+  private final String message;
+  
+  public PrivateMessageEvent(Player sender, Player recipient, String message) {
+    this.sender = sender;
+    this.recipient = recipient;
+    this.message = message;
+  }
+  
+  public Player getSender() {
+    return sender;
+  }
+  
+  public Player getRecipient() {
+    return recipient;
+  }
+  
+  public String getMessage() {
+    return message;
+  }
+  
+  // toString, equals, and hashCode may be added as needed
+
+}
+
+```
+
+You'll notice that your events don't need to extend or implement anything. They just work.
+
+### Firing the Event
+
+To fire the event, you'll need to get the server's event manager and use the `fire` method. Note that this returns a `CompletableFuture`, so if you want to continue logic after the event is handled by all listeners, use a callback:
+
+```java
+server.getEventManager().fire(new PrivateMessageEvent(sender, recipient, message)).thenAccept((event) -> {
+  // event has finished firing
+  // do some logic dependent on the result
+});
+```
+
+### Adding event state
+
+Velocity uses the generalised `ResultedEvent` for events which have some sort of 'result'.
```

Comment:
We should give `ResultedEvent` a better treatment - it's a fundamental API to the Velocity ecosystem. There may be some historical sources you can dig up that explain why this design decision was made.